### PR TITLE
refactor: unify all forums into a single XenForo crawler

### PR DESCRIPTION
**Pull request by NTFSvolume**
Defines a single `XenforoCrawler` to deduce code duplication. The code on every crawler was basically the same

However, given that each forum uses slightly different CSS selectors, using a single crawler to manage all of them will make it really annoying to work with. What i did instead was create a base `XenforoCrawler` and each forum inherit from it, overriding key properties like the `primary_base_domain` and the CSS selectors they use

### CODE REVIEW

**Comment by jbsparrow** [comment]
It's beautiful 😻

**Comment by NTFSvolume** [comment]
Tested on SMG and SC. Everything seems to be working

**Comment by NTFSvolume** [comment]
Added a fix to the chevereto profile scraper. It’s unrelated to this PR, but I found it while testing and it’s a simple one-line fix.

**Review by jbsparrow — APPROVED** [review]
I don't really use any of the forums except SimpCity but I'll look into making accounts and testing the others before we merge.

**Comment by jbsparrow** [comment]
- [ ] Test Celebforum
- [ ] Test F95zone
- [ ] Test Leakedmodels
- [ ] Test Nudostar
- [ ] Test Xbunker
- [ ] Test SimpCity
- [ ] Test Socialmediagirls

File: cyberdrop_dl/scraper/crawlers/f95zone_crawler.py [hunk 1]
```diff
@@ -1,364 +1,21 @@
...
-    from bs4 import BeautifulSoup, Tag
-
     from cyberdrop_dl.managers.manager import Manager
 
 
-class F95ZoneCrawler(Crawler):
+class F95ZoneCrawler(XenforoCrawler):
+    primary_base_domain = URL("https://f95zone.to")
+    post_selectors = PostSelectors(
+        date=Selector("time", "data-time"),
```
Comment (jbsparrow) [inline cyberdrop_dl/scraper/crawlers/f95zone_crawler.py:16]:
```suggestion
        date=Selector("time", "data-time"),
        number=Selector("a[class=u-concealed]", "href"),
```

File: cyberdrop_dl/scraper/crawlers/nudostar_crawler.py [hunk 2]
```diff
@@ -1,355 +1,21 @@
...
-    from bs4 import BeautifulSoup, Tag
-
     from cyberdrop_dl.managers.manager import Manager
 
 
-class NudoStarCrawler(Crawler):
+class NudoStarCrawler(XenforoCrawler):
+    primary_base_domain = URL("https://nudostar.com")
+    post_selectors = PostSelectors(
+        date=Selector("time", "data-time"),
```
Comment (jbsparrow) [inline cyberdrop_dl/scraper/crawlers/nudostar_crawler.py:16]:
```suggestion
        date=Selector("time", "data-time"),
        number=Selector("a[class=u-concealed]", "href"),
```

File: cyberdrop_dl/scraper/crawlers/xenforo_crawler.py [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+class XenforoCrawler(Crawler):
+    login_required = True
+    primary_base_domain = None
+    selectors = XenforoSelectors()
+
+    def __init__(self, manager: Manager, site: str, folder_domain: str | None = None) -> None:
+        super().__init__(manager, site, folder_domain)
+        self.primary_base_domain = self.primary_base_domain or URL(f"https://{site}")
+        self.attachment_url_part = ["attachments"]
+        self.attachment_url_hosts = ["smgmedia"]
```
Comment (jbsparrow) [inline cyberdrop_dl/scraper/crawlers/xenforo_crawler.py:92]:
```suggestion
        self.attachment_url_hosts = ["smgmedia", "attachments.f95zone"]
```

**Review by jbsparrow — CHANGES_REQUESTED** [review]
Now F95zone is the only forum that has some issues. With the selector I added for f95zone attachments, we also grab thumbnails. F95zone also has a redirect protection thing, we can easily get the correct URL from it but I'm not sure exactly where in the code we would put the extractor for it.

**Comment by NTFSvolume** [comment]
I think i know what the issue is. F95zone had a different logic for `handle_confirmation_link` . I renamed that function to `handle_link_confirmation2` but it's never used in the current crawler cause i didn't know why it was different. 

We can use that one as an override of the base one in `F95zoneCrawler`.

https://github.com/jbsparrow/CyberDropDownloader/blob/95a6e0662a320dea066776353ad1d86dff993ac3/cyberdrop_dl/scraper/crawlers/xenforo_crawler.py#L303-L314

**Comment by NTFSvolume** [comment]
Pushed a commit with the override. Please try again with F95Zone

**Comment by NTFSvolume** [comment]
The old F95Zone called `handle_link_confirmation` if the link had the word `masked`, but i think that's too generic. A regular URL could have that word as well. Can you confirm what the link looks like? maybe `masked` is a part of the path and we can limit the condition a bit more

**Comment by NTFSvolume** [comment]
Also, the link confirmation override makes a random call to get the soup of the link but the result is not saved anywhere. Pretty sure is not necessary, but I kept it as is to match the original logic.

**Comment by jbsparrow** [comment]
For handling the confirmation we can just skip the redirect by extracting the target url from the redirect url. Wouldn't that be more ideal?

**Review by jbsparrow — APPROVED** [review]

**Comment by jbsparrow** [comment]
Fixed F95zone and added a function for checking if a URL is a confirmation link so that we can define it per-crawler.

**Pull request merged**